### PR TITLE
Put log entries into HOBs in PEI during log buffer initialization

### PR DESCRIPTION
**Pull request by kuqin12**
# Preface

Please ensure you have read the [contribution docs](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md) prior
to submitting the pull request. In particular,
[pull request guidelines](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md#pull-request-best-practices).

## Description

The current implementation will directly allocation pages through PEI services when Advanced logger is not initialized. This cause an issue of permanent loop if/when there are prints from the allocation routine.

This change will create a hob during this interim state and log messages to hobs temporarily and coalesce the hobs after the allocation is finalized. This will reduce the print restriction down to hob creation only.

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

This is tested on QEMU SBSA as well as Q35. Both booted to UEFI shell.

## Integration Instructions

N/A

### CODE REVIEW

File: AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c [hunk 1]
```diff
@@ -492,6 +499,35 @@ AdvancedLoggerGetLoggerInfo (
...
+  GuidHobInterim = GetFirstGuidHob (&gAdvancedLoggerInterimHobGuid);
+  if (GuidHobInterim != NULL) {
+    // In the middle of initialization, save the log to the interim hobs
+    Status = PeiServicesCreateHob (
+               EFI_HOB_TYPE_GUID_EXTENSION,
+               (UINT16)(sizeof (EFI_HOB_GUID_TYPE) + sizeof (ADVANCED_LOGGER_INFO) + ADVANCED_LOGGER_MAX_MESSAGE_SIZE),
+               (VOID **)&GuidHobInterimBuf
+               );
+    LoggerInfo = (ADVANCED_LOGGER_INFO *)GET_GUID_HOB_DATA (GuidHobInterimBuf);
+    ZeroMem ((VOID *)LoggerInfo, BufferSize);
```
Comment (os-d) [inline AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c:516]: I'm missing where BufferSize is getting set?

File: AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c [hunk 2]
```diff
@@ -516,14 +552,16 @@ AdvancedLoggerGetLoggerInfo (
 
       if (FeaturePcdGet (PcdAdvancedLoggerPeiInRAM)) {
         Pages =  FixedPcdGet32 (PcdAdvancedLoggerPages);
+        Type  =  EfiReservedMemoryType;
       } else {
         Pages =  FixedPcdGet32 (PcdAdvancedLoggerPreMemPages);
+        Type  =  EfiBootServicesData;
```
Comment (os-d) [inline AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c]: Probably worth a comment that this is to avoid the 64k issue.

**Review by os-d — APPROVED** [review]

File: AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c (hunk 1 above)
Comment (kuqin12) [reply AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c:516]: Good catch! Fixed.

File: AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c (hunk 2 above)
Comment (kuqin12) [reply AdvLoggerPkg/Library/AdvancedLoggerLib/PeiCore/AdvancedLoggerLib.c]: Updated. Thanks~

**Review by apop5 — APPROVED** [review]

**Pull request merged**